[RDY] win: Fix reading content from stdin

b-r-o-c-k opened the pull request:
Fixes #6890

justinmk commented:
Wonderful, this should allow us to nudge https://github.com/neovim/neovim/pull/6299

janlazo commented:
@b-r-o-c-k I needed this https://github.com/neovim/neovim/pull/8160/commits/bc43d2559f33aa6334d70834b389f0bc59dadcbf for my PR since your workaround for lastexitcode didn't work for MINGW (or something else is happening). Does it need stderr redirected to stdout?

justinmk commented on src/nvim/fileio.c:
```diff
@@ -1737,9 +1737,17 @@ readfile (
     close(0);
+#ifndef WIN32
+    // On Unix, use stderr for stdin, makes shell commands work.
     ignored = dup(2);
```
In Vim this is guarded by `HAVE_DUP`, which is defined for Windows (`src/os_dos.h` in Vim's source).

https://github.com/vim/vim/blob/e7499ddc33508d3d341e96f84a0e7b95b2d6927c/src/fileio.c#L2397-L2398

I wonder why it doesn't work in Nvim.

b-r-o-c-k commented on src/nvim/tui/input.c:
```diff
@@ -47,7 +47,13 @@ void term_input_init(TermInput *input, Loop *loop)
+                            FILE_SHARE_READ, (LPSECURITY_ATTRIBUTES)NULL,
+                            OPEN_EXISTING, 0, (HANDLE)NULL);
+  const int conin_fd = _open_osfhandle(conin_handle, _O_RDONLY);
+  uv_tty_init(loop, &input->tty_in, conin_fd, true);
```
I think it's a better idea to move this patch to here because it's going to be less redundant. The patch won't need to be duplicated in #6299 if it's moved to here.

justinmk replied:
@b-r-o-c-k seems reasonable, but then it won't work for `--headless` (TUI is not started then). So some duplication or refactoring might still be needed.

b-r-o-c-k replied:
You're right, this doesn't work for headless mode and I'm not sure if there's a good way to avoid the duplication so I'll just change it back to the way it was.

justinmk commented:
can you include your explanation from https://github.com/neovim/neovim/pull/8267#discussion_r181260964 in the commit message?

b-r-o-c-k committed:
win: Fix reading content from stdin

Fixes #6890 by reading from the Windows console input buffer after
stdin has been closed. Vim normally reads from this buffer when it
fails to read from stdin:
https://github.com/vim/vim/blob/e7499ddc33508d3d341e96f84a0e7b95b2d6927c/src/os_win32.c#L1703-L1714
